slick/src/main/scala/slick/jdbc/MySQLProfile.scala

```diff
@@ -252,14 +252,12 @@ trait MySQLProfile extends JdbcProfile { profile =>
   }
 
   class UpsertBuilder(ins: Insert) extends super.UpsertBuilder(ins) {
+    @deprecated("Method kept for binary compatibility in 3.x, but should be removed in 3.4.x")
     def buildInsertIgnoreStart: String = allNames.iterator.mkString(s"insert ignore into $tableName (", ",", ") ")
     override def buildInsert: InsertBuilderResult = {
-      val start = buildInsertIgnoreStart
-      if (softNames.isEmpty) new InsertBuilderResult(table, s"$start values $allVars", syms)
-      else {
-        val update = softNames.map(n => s"$n=VALUES($n)").mkString(", ")
-        new InsertBuilderResult(table, s"$start values $allVars on duplicate key update $update", syms)
-      }
+      val start = buildInsertStart
+      val update = allNames.map(n => s"$n=VALUES($n)").mkString(", ")
+      new InsertBuilderResult(table, s"$start values $allVars on duplicate key update $update", syms)
```

Comment:
I haven't dug into the code. Is `allNames` guaranteed not to be empty? What exactly are `softNames` and `allNames`?

Reply:
> I haven't dug into the code. Is allNames guaranteed not to be empty?

As far as I know there is no API to specify an empty column name list. 

> What exactly are softNames and allNames?

- `allNames` is all columns in the table.
https://github.com/slick/slick/blob/98ded3a777372d0783dbb42233b08ff51f6da3a7/slick/src/main/scala/slick/jdbc/JdbcStatementBuilderComponent.scala#L501-L502

- `softNames` is all columns in the table excludes PK.
https://github.com/slick/slick/blob/98ded3a777372d0783dbb42233b08ff51f6da3a7/slick/src/main/scala/slick/jdbc/JdbcStatementBuilderComponent.scala#L542-L546